mlir/lib/Transforms/Utils/WalkPatternRewriteDriver.cpp

```diff
@@ -0,0 +1,86 @@
+//===- WalkPatternRewriteDriver.cpp - A fast walk-based rewriter ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Implements mlir::walkAndApplyPatterns.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Transforms/WalkPatternRewriteDriver.h"
+
+#include "mlir/IR/OperationSupport.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Verifier.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/ErrorHandling.h"
+
+#define DEBUG_TYPE "walk-rewriter"
+
+namespace mlir {
+
+namespace {
+// Forwarding listener to guard against unsupported erasures. Because we use
+// walk-based pattern application, erasing the op from the *next* iteration
+// (e.g., a user of the visited op) is not valid.
+struct ErasedOpsListener final : RewriterBase::ForwardingListener {
+  using RewriterBase::ForwardingListener::ForwardingListener;
+
+  void notifyOperationErased(Operation *op) override {
+    if (op != visitedOp)
+      llvm::report_fatal_error("unsupported op erased in WalkPatternRewriter; "
+                               "erasure is only supported for matched ops");
```

Comment:
Seems like this could be an assert (or llvm_unreachable)?

Reply:
Note that this is only used under `MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS`. Greedy pattern rewriter also uses `report_fatal_error` for [similar checks](https://github.com/llvm/llvm-project/blob/7b3da7b3b2b0e2f322dddf1f343571cc7fd09b09/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp#L83-L88) -- I tried to follow that convention here

Reply:
This is motivating using llvm_unreachable actually: this is a virtual method that will be dead-code in production (any build without MLIR_ENABLE_EXPENSIVE_PATTERN_API_CHECKS), but the compiler can't optimize it away because it is a virtual method.

Reply:
I added ifdefs around the whole class to make it clear that this won't be used in production and won't generate any extra code.